src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt

```diff
@@ -85,6 +85,30 @@ class JUnitTestsAssembler(
         log.debug(super.getContent())
     }
 
+    /**
+     * Receives a response from the HuggingFace API and processes it.
+     */
+    fun consumeHFRequest(httpRequest: HttpRequests.Request) {
+        Thread.sleep(50L)
+        val text = httpRequest.reader.readLine()
+        val generatedTestCases = removeDelimiterLines(
+            JsonParser.parseString(text).asJsonArray[0]
+                .asJsonObject["generated_text"].asString.trim(),
+        )
+        super.consume(generatedTestCases)
+        log.debug(super.getContent())
+        updateProgressBar()
+    }
+
+    /**
+     * Removes delimiter lines used for code blocks in LLMs' response
+     */
+    private fun removeDelimiterLines(text: String): String {
+        return text.lines()
+            .filter { line -> line != "```" && line != "```java" }
+            .joinToString("\n")
+    }
```

Comment:
Do I understand correctly that this function serves merely as a "hack" to avoid empty triple-tick blocks? 

As I see from the HF request body, oftentimes, the LLM's response contains an empty block of ` ``` `, and only after provides a test suite (the selected model is `Meta-Llama-3-8B-Instruct`).

Reply:
The function extracts generated test suites from the LLM's response.

Reply:
There is an oversight in this implementation. Let me explain by the example test suite generated by `Meta-Llama-3-8B-Instruct`:

Having the following prompt:
````log
<s>[INST] <<SYS>> You are a helpful and honest code and programming assistant. Please, respond concisely and truthfully. <</SYS>> Generate unit tests in Java for Calc.
All tests should be for JUnit 4.
All test cases MUST be inserted into a single test suite.
Name all methods according to the template - [MethodUnderTest][Scenario]Test, and use only English letters.
The source code of class under test is as follows:
```
public class Calc {
    public int sum(int a, int b) {
        return a + b;
    }

    public int multiply(int a, int b) {
        return a * b;
    }
}
```

Generate public classes and public methods. Response only a code with tests between ```, do not provide any other text. [/INST]
````
The LLM responds with the following output:
````log
```

```java
```java
import org.junit.Before;
import org.junit.Test;
import static org.junit.Assert.assertEquals;

public class CalcTest {
    private Calc calc;

    @Before
    public void setup() {
        calc = new Calc();
    }

    @Test
    public void sumTwoPositiveNumbersTest() {
        assertEquals(4, calc.sum(2, 2));
    }

    @Test
    public void sumTwoNegativeNumbersTest() {
        assertEquals(-4, calc.sum(-2, -2));
    }

    @Test
    public void sumPositiveAndNegativeNumberTest() {
        assertEquals(0, calc.sum(2, -2));
    }

    @Test
    public void multiplyTwoPositiveNumbersTest() {
        assertEquals(6, calc.multiply(2, 3));
    }

    @Test
    public void multiplyTwoNegativeNumbersTest() {
        assertEquals(-6, calc.multiply(-2, -3));
    }

    @Test
    public void multiplyPositiveAndNegativeNumberTest() {
        assertEquals(-6, calc.multiply(2, -3));
    }
}
```java``````java```
```java``````java```
```java``````java```
````
The problem arises for the trailing triple-tick blocks at the end: The condition within `filter` of `removeDelimiterLines` is true for them, as the tick blocks appear on the same line (therefore, the `line` is neither \`\`\`-block nor \`\`\`java-block).

### Why it breaks things:

Our parser `org.jetbrains.research.testspark.core.test.parsers.java.JUnitTestSuiteParser` has a method `parseTestSuite` which is responsible for parsing the test suite from the text into a `TestSuiteGeneratedByLLM` DTO structure, and it has the following code:
```java
var rawCode = rawText

if (rawText.contains("```")) {
    rawCode = rawText.split("```")[1]
}
```
This piece of code on the above LLM response will yield the string `"java"`, not the test suite text, therefore the test suite is lost.

Responses alike the above are very common, if not normal, at least to `Meta-Llama-3-8B-Instruct` (maybe for the 2nd model it is different; cannot test without the paid subscription).

### How to fix
These empty triple-tick blocks should be sieved out from the response; it may be done as follows:
```kotlin
private fun removeDelimiterLines(text: String): String {
    return text
        .replace("```java", "```") // making closing tick-blocks be the same as the opening ones
        .split("```") // splitting into potential code chunks; we assume that only one contains the test suite
        .filter { it.isNotBlank() } // remove empty code chunks (assuming all of them are empty but one with the test suite)
        .joinToString { codeChunk -> "```\n$codeChunk\n```" } // embrace the code chunk with the test suite into tick-block, so it can be parsed by `JUnitTestSuiteParser` further
}
```
Perhaps, the proposal might not cover some edge-case responses, but even as is it significantly increases the success rate making the LLM usable.

Reply:
I've implemented a similar fix in https://github.com/JetBrains-Research/TestSpark/pull/212/commits/83bdd2be0577de49cd277f18755be7b6eb9a2fa7 to make the LLM response compatible with the parser.